consensus/src/liveness/proposal_generator.rs

```diff
@@ -285,10 +300,40 @@ impl ProposalGenerator {
 
             let voting_power_ratio = proposer_election.get_voting_power_participation_ratio(round);
 
-            let (max_block_txns, max_block_bytes, proposal_delay) = self
+            let (mut max_block_txns, mut max_block_bytes, proposal_delay) = self
                 .calculate_max_block_sizes(voting_power_ratio, timestamp)
                 .await;
 
+            let sys_txns = if self.should_propose_sys_txns {
+                let pending_sys_txns: HashSet<HashValue> = pending_blocks
+                    .iter()
+                    .filter_map(|block| block.sys_txns())
+                    .flatten()
+                    .map(SystemTransaction::hash)
+                    .collect();
+
+                // Priority: self.sys_txn_provider[0] > ... > self.sys_txn_provider[-1] > payload.
+                let mut sys_txns = vec![];
+                for provider in self.sys_txn_providers.iter() {
```

Comment:
Do we want to have a separate limit for system txns?

Reply:
I'm not sure. but IMO we don't need it for now and can add it when it's necessary.

Reply:
@zekun000 @grao1991 WDYT?